### ISSUE TITLE
Out of memory detected with kafka-connect-elasticsearch 10.0.0

### ISSUE THREAD

**Issue by pkgonan** [confluentinc/kafka-connect-elasticsearch#444]
Hi, memory leak is detected at kafka-connect-elasticsearch version 10.0.0.

Before 18:33 - Deployment version : kafka-connect-elasticsearch 10.0.0
After 18:33 - Deployment version : kafka-connect-elasticsearch 5.5.1

![Old Gen](https://user-images.githubusercontent.com/15858772/94417891-a1343f00-01bb-11eb-9c7f-feb90b4ee055.png)

[Before kafka-connect-elasticsearch 10.0.0 - Dockerfile]
```
FROM confluentinc/cp-kafka-connect-base:6.0.0

ENV CONNECT_PLUGIN_PATH="/usr/share/java,/usr/share/confluent-hub-components" \
    CUSTOM_SMT_PATH="/usr/share/confluent-hub-components/custom-smt"


ARG DEBEZIUM_VERSION=1.2.5
ARG CONNECT_TRANSFORM_VERSION=1.3.2
ARG CONNECT_ELASTICSEARCH_VERSION=10.0.0


# Download Using confluent-hub
RUN confluent-hub install --no-prompt confluentinc/connect-transforms:$CONNECT_TRANSFORM_VERSION \
 && confluent-hub install --no-prompt confluentinc/kafka-connect-elasticsearch:$CONNECT_ELASTICSEARCH_VERSION


# Download Custom SMT
RUN mkdir $CUSTOM_SMT_PATH && cd $CUSTOM_SMT_PATH && \
    curl -sO https://repo1.maven.org/maven2/io/debezium/debezium-core/$DEBEZIUM_VERSION.Final/debezium-core-$DEBEZIUM_VERSION.Final.jar
```

[After kafka-connect-elasticsearch 5.5.1 - Dockerfile]
```
FROM confluentinc/cp-kafka-connect-base:6.0.0

ENV CONNECT_PLUGIN_PATH="/usr/share/java,/usr/share/confluent-hub-components" \
    CUSTOM_SMT_PATH="/usr/share/confluent-hub-components/custom-smt"


ARG DEBEZIUM_VERSION=1.2.5
ARG CONNECT_TRANSFORM_VERSION=1.3.2
ARG CONNECT_ELASTICSEARCH_VERSION=5.5.1


# Download Using confluent-hub
RUN confluent-hub install --no-prompt confluentinc/connect-transforms:$CONNECT_TRANSFORM_VERSION \
 && confluent-hub install --no-prompt confluentinc/kafka-connect-elasticsearch:$CONNECT_ELASTICSEARCH_VERSION


# Download Custom SMT
RUN mkdir $CUSTOM_SMT_PATH && cd $CUSTOM_SMT_PATH && \
    curl -sO https://repo1.maven.org/maven2/io/debezium/debezium-core/$DEBEZIUM_VERSION.Final/debezium-core-$DEBEZIUM_VERSION.Final.jar
```

[Heap Dump Analyze - 1]
![Heap Dump](https://user-images.githubusercontent.com/15858772/94418594-90d09400-01bc-11eb-8ae2-ec8659056cbc.png)

[Heap Dump Analyze - 2]
![Heap Dump 분석 1](https://user-images.githubusercontent.com/15858772/94418642-9fb74680-01bc-11eb-9d0e-60b0aa9ae6e1.png)

[Heap Dump Analyze - 3]
![Heap Dump 분석 2](https://user-images.githubusercontent.com/15858772/94418669-a940ae80-01bc-11eb-8c41-8c9a78d0d18c.png)

[Heap Dump Analyze - 4]
![Heap Dump 분석 3](https://user-images.githubusercontent.com/15858772/94418685-acd43580-01bc-11eb-94c5-8ef44a53a676.png)

**Comment by levzem** [comment]
@pkgonan I have a hunch on where this potential leak is. Can you post your connector configs?

**Comment by levzem** [comment]
@pkgonan also do you have an approximation of your message size?

**Comment by pkgonan** [comment]
@levzem hi
**I registered 16 connectors same format for extracting primary key.**

[primary key : id - only has tasks.max - 10]
```
    "connector.class": "io.confluent.connect.elasticsearch.ElasticsearchSinkConnector",
    "tasks.max": "10",
    "topics": "a, b, c, d, e, f, g, .... about 50 topics"
    "connection.url": "'${elasticsearch_url}'",
    "transforms": "unwrap,extract",
    "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
    "transforms.unwrap.drop.tombstones": "false",
    "transforms.extract.type": "org.apache.kafka.connect.transforms.ExtractField$Key",
    "transforms.extract.field": "id",
    "connection.compression": "true",
    "linger.ms": "50",
    "batch.size": "500",
    "max.buffered.records": "1000",
    "max.retries": "10",
    "key.ignore": "false",
    "behavior.on.null.values": "delete",
    "behavior.on.malformed.documents": "warn",
    "errors.tolerance": "all",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "type.name": "_doc",
    "drop.invalid.message": "true"
```

[primary key : appleId]
```
    "connector.class": "io.confluent.connect.elasticsearch.ElasticsearchSinkConnector",
    "tasks.max": "3",
    "topics": "AppleLogin",
    "connection.url": "'${elasticsearch_url}'",
    "transforms": "unwrap,extract",
    "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
    "transforms.unwrap.drop.tombstones": "false",
    "transforms.extract.type": "org.apache.kafka.connect.transforms.ExtractField$Key",
    "transforms.extract.field": "appleId",
    "connection.compression": "true",
    "linger.ms": "50",
    "batch.size": "500",
    "max.buffered.records": "1000",
    "max.retries": "10",
    "key.ignore": "false",
    "behavior.on.null.values": "delete",
    "behavior.on.malformed.documents": "warn",
    "errors.tolerance": "all",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "type.name": "_doc",
    "drop.invalid.message": "true"
```

[primary key : googleId]
```
    "connector.class": "io.confluent.connect.elasticsearch.ElasticsearchSinkConnector",
    "tasks.max": "3",
    "topics": "cdc-data.azar.'${database_name}'.GoogleLogin",
    "connection.url": "'${elasticsearch_url}'",
    "transforms": "unwrap,extract",
    "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
    "transforms.unwrap.drop.tombstones": "false",
    "transforms.extract.type": "org.apache.kafka.connect.transforms.ExtractField$Key",
    "transforms.extract.field": "googleId",
    "connection.compression": "true",
    "linger.ms": "50",
    "batch.size": "500",
    "max.buffered.records": "1000",
    "max.retries": "10",
    "key.ignore": "false",
    "behavior.on.null.values": "delete",
    "behavior.on.malformed.documents": "warn",
    "errors.tolerance": "all",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "type.name": "_doc",
    "drop.invalid.message": "true"
```

[primary key : userId]
```
    "connector.class": "io.confluent.connect.elasticsearch.ElasticsearchSinkConnector",
    "tasks.max": "3",
    "topics": "User",
    "connection.url": "'${elasticsearch_url}'",
    "transforms": "unwrap,extract",
    "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
    "transforms.unwrap.drop.tombstones": "false",
    "transforms.extract.type": "org.apache.kafka.connect.transforms.ExtractField$Key",
    "transforms.extract.field": "userId",
    "connection.compression": "true",
    "linger.ms": "50",
    "batch.size": "500",
    "max.buffered.records": "1000",
    "max.retries": "10",
    "key.ignore": "false",
    "behavior.on.null.values": "delete",
    "behavior.on.malformed.documents": "warn",
    "errors.tolerance": "all",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "errors.log.enable": "true",
    "errors.log.include.messages": "true",
    "type.name": "_doc",
    "drop.invalid.message": "true"
```

...  and other connectors same format.

**Comment by levzem** [comment]
@pkgonan thanks for the update!

I have a pending PR with the fix #446 that we should have out and released soon

**Comment by pkgonan** [comment]
@levzem 
Thank you!
If #446 PR released, can I download it immediately from [Confluent Hub - kafka-connect-elasticsearch](https://www.confluent.io/hub/confluentinc/kafka-connect-elasticsearch) ?

**Comment by levzem** [comment]
we will issue a `10.0.1` bug fix release to be available on the hub

**Comment by levzem** [comment]
@pkgonan, `10.0.1` was just released including #446 fix for the OOM you were seeing, therefore closing

**Comment by pkgonan** [comment]
@levzem 
Thank you!
